File: rednote/get-qrcode/scripts/get-qrcode.js
#!/usr/bin/env node

const path = require('path');
const fs = require('fs');

const libPath = path.join(__dirname, '..', 'lib');
const { BrowserManager, getBrowserManager, closeAll } = require(path.join(libPath, 'browser'));
const { CookieManager, syncCookiesFromContext } = require(path.join(libPath, 'cookie'));
const { logger, randomSleep, formatDate } = require(path.join(libPath, 'utils'));

const DATA_PATH = process.env.XHS_DATA_PATH || path.join(__dirname, '..', '..', 'data');
const CREATOR_URL = 'https://creator.xiaohongshu.com/';
const LOGIN_URL = 'https://creator.xiaohongshu.com/login';

const SELECTORS = {
  qrcode: '.qrcode-box img, .login-qrcode img, [class*="qr"] img',
  qrcodeCanvas: '.qrcode-box canvas, .login-qrcode canvas',
  userInfo: '.user-info',
  userName: '.user-name',
};

async function getQrcode(options = {}) {
  const dataPath = options.dataPath || DATA_PATH;
  const savePath = options.savePath || path.join(dataPath, 'qrcode.png');
  
  logger.info('获取登录二维码...');
  logger.info(`数据路径: ${dataPath}`);
  
  const browserManager = getBrowserManager(dataPath, { headless: false });
  const cookieManager = new CookieManager(dataPath);
  
  cookieManager.ensureDataDir();
  
  try {
    const page = await browserManager.getPage();
    
    logger.info('访问登录页面...');
    await page.goto(CREATOR_URL, { waitUntil: 'domcontentloaded' });
    
    await randomSleep(2000, 3000);
    
    const qrcodeImg = await page.$(SELECTORS.qrcode);
    if (qrcodeImg) {
      logger.info('找到二维码图片');
      
      await qrcodeImg.screenshot({ path: savePath });
      logger.info(`二维码已保存: ${savePath}`);
      
      return {
        success: true,
        qrcodePath: savePath,
        message: '请扫描二维码登录',
        deadline: new Date(Date.now() + 180000).toISOString(),
        browserManager,
        page,
      };
    }
    
    const qrcodeCanvas = await page.$(SELECTORS.qrcodeCanvas);
    if (qrcodeCanvas) {
      logger.info('找到二维码 Canvas');
      
      const buffer = await qrcodeCanvas.screenshot();
      fs.writeFileSync(savePath, buffer);
      logger.info(`二维码已保存: ${savePath}`);
      
      return {
        success: true,
        qrcodePath: savePath,
        message: '请扫描二维码登录',
        deadline: new Date(Date.now() + 180000).toISOString(),
        browserManager,
        page,
      };
    }
    
    const pageUrl = page.url();
    if (pageUrl.includes('/creator.xiaohongshu.com') && !pageUrl.includes('/login')) {
      logger.info('已处于登录状态');
      
      return {
        success: true,
        isLoggedIn: true,
        message: '已登录，无需获取二维码',
        browserManager,
        page,
      };
    }
    
    logger.warn('未找到二维码元素，尝试截取整个页面');
    
    const fullSavePath = path.join(dataPath, 'login-page.png');
    await page.screenshot({ path: fullSavePath, fullPage: false });
    
    return {
      success: true,
      qrcodePath: fullSavePath,
      message: '请查看登录页面截图',
      deadline: new Date(Date.now() + 180000).toISOString(),
      browserManager,
      page,
    };
    
  } catch (error) {
    logger.error('获取二维码失败:', error.message);
    
    return {
      success: false,
      error: error.message,
    };
  }
}

async function waitForLogin(options = {}) {
  const timeout = (options.timeout || 120) * 1000;
  const dataPath = options.dataPath || DATA_PATH;
  const cookieManager = new CookieManager(dataPath);
  
  logger.info('等待扫码登录...');
  logger.info(`超时时间: ${timeout / 1000} 秒`);
  
  const result = await getQrcode(options);
  
  if (!result.success) {
    return result;
  }
  
  if (result.isLoggedIn) {
    return result;
  }
  
  console.log('\n' + '='.repeat(50));
  console.log('📱 请用小红书 App 扫描二维码登录');
  console.log(`   二维码路径: ${result.qrcodePath}`);
  console.log(`   有效期: 3 分钟`);
  console.log('='.repeat(50) + '\n');
  
  const page = result.page;
  const startTime = Date.now();
  
  while (Date.now() - startTime < timeout) {
    await randomSleep(2000, 3000);
    
    const currentUrl = page.url();
    
    if (currentUrl.includes('/creator.xiaohongshu.com') && !currentUrl.includes('/login')) {
      logger.info('检测到页面跳转，登录成功');
      
      await randomSleep(2000, 3000);
      
      const userInfo = await page.$(SELECTORS.userInfo);
      if (userInfo) {
        let username = '';
        let userId = '';
        
        try {
          const userNameEl = await page.$(SELECTORS.userName);
          if (userNameEl) {
            username = await userNameEl.textContent() || '';
            username = username.trim();
          }
        } catch (e) {
          logger.warn('获取用户名失败:', e.message);
        }
        
        await syncCookiesFromContext(result.browserManager.getContext(), cookieManager);
        
        await cookieManager.saveLoginState({
          isLoggedIn: true,
          username,
          userId,
        });
        
        console.log('\n' + '='.repeat(50));
        console.log('✅ 登录成功！');
        console.log(`用户名: ${username || '未知'}`);
        console.log('='.repeat(50) + '\n');
        
        return {
          success: true,
          isLoggedIn: true,
          username,
          userId,
          message: '登录成功',
        };
      }
    }
    
    const elapsed = Math.floor((Date.now() - startTime) / 1000);
    if (elapsed % 10 === 0) {
      logger.info(`等待扫码登录... (${elapsed}秒)`);
    }
  }
  
  logger.warn('等待登录超时');
  
  return {
    success: false,
    isLoggedIn: false,
    message: '等待登录超时，请重新获取二维码',
  };
}

async function main() {
  const args = process.argv.slice(2);
  const noWait = args.includes('--no-wait');
  
  let result;
  
  if (noWait) {
    result = await getQrcode();
    
    if (result.success) {
      console.log('\n' + '='.repeat(50));
      console.log('📱 登录二维码');
      console.log(`   路径: ${result.qrcodePath}`);
      console.log(`   有效期: 3 分钟`);
      console.log('='.repeat(50) + '\n');
    }
  } else {
    result = await waitForLogin();
  }
  
  if (!result.success) {
    console.log('\n❌ 失败:', result.message || result.error);
    process.exit(1);
  }
  
  process.exit(0);
}

if (require.main === module) {
  main().catch(error => {
    console.error('执行失败:', error);
    process.exit(1);
  });
}

module.exports = { getQrcode, waitForLogin };

File: .claude/skills/power-media/rednote/get-qrcode/scripts/get_qrcode.py
#!/usr/bin/env python3
"""
获取小红书创作者平台登录二维码
"""
import sys
import os
import json
import time

# 添加 lib 到路径
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '../../lib'))
from rednote_automation import RedNoteAutomation


def wait_for_login(automation, timeout=120, check_interval=3):
    """
    等待用户扫码登录
    
    参数:
        automation: RedNoteAutomation 实例
        timeout: 超时时间（秒），默认 120 秒
        check_interval: 检查间隔（秒），默认 3 秒
    """
    print(f"\n等待用户扫码登录（最多等待 {timeout} 秒）...")
    start_time = time.time()
    check_count = 0

    while time.time() - start_time < timeout:
        check_count += 1
        elapsed = int(time.time() - start_time)
        print(f"\n[{elapsed}s] 第 {check_count} 次检查登录状态...")

        # 截图检查
        result = automation.check_login_status()

        if not result.get("success"):
            print(f"  截图失败: {result.get('error')}")
            time.sleep(check_interval)
            continue

        screenshot_path = result.get("screenshot_path")
        print(f"  截图已保存: {screenshot_path}")
        print(f"  请 AI 分析截图判断是否已登录")

        # AI 需要分析截图，如果检测到已登录（有用户头像），则退出循环
        # 这里我们返回截图信息，由 AI 判断
        # 实际使用中，AI 会分析截图并告诉脚本是否已登录

        # 暂时返回，等待 AI 分析
        return {
            "success": True,
            "status": "waiting_for_ai_analysis",
            "screenshot_path": screenshot_path,
            "elapsed": elapsed,
            "message": "请 AI 分析截图判断是否已登录"
        }

        time.sleep(check_interval)

    # 超时
    return {
        "success": False,
        "status": "timeout",
        "message": f"登录超时（{timeout} 秒），请刷新页面重新获取二维码",
        "elapsed": timeout
    }


def main():
    """主函数"""
    print("=" * 60)
    print("📱 获取小红书创作者平台登录二维码")
    print("=" * 60)
    print()

    automation = RedNoteAutomation()

    # 1. 查找/打开窗口
    print("1. 查找创作者平台窗口...")
    if not automation.find_or_open_creator():
        print("✗ 无法打开创作者平台窗口")
        print(json.dumps({
            "success": False,
            "error": "无法打开创作者平台窗口"
        }, ensure_ascii=False, indent=2))
        return

    # 2. 检查是否已登录
    print("\n2. 检查当前登录状态...")
    login_status = automation.check_login_status()

    if not login_status.get("success"):
        print(f"✗ 检查失败: {login_status.get('error')}")
        print(json.dumps(login_status, ensure_ascii=False, indent=2))
        return

    screenshot_path = login_status.get("screenshot_path")
    print(f"\n✓ 截图已保存: {screenshot_path}")
    print("\n请 AI 分析截图:")
    print("  - 如已登录（有用户头像）→ 提示用户已登录，无需扫码")
    print("  - 如未登录（显示二维码）→ 提示用户扫码")

    # 3. 等待登录（如未登录）
    print("\n3. 等待用户扫码登录...")
    print("   请使用小红书 APP 扫描屏幕上的二维码")
    print("   二维码有效期约 2 分钟，超时需刷新页面\n")

    result = wait_for_login(automation, timeout=120, check_interval=3)

    print("\n" + "=" * 60)
    print("JSON 输出:")
    print("=" * 60)
    print(json.dumps(result, ensure_ascii=False, indent=2))


if __name__ == "__main__":
    main()
